test/middleware/express.test.ts

```diff
@@ -0,0 +1,341 @@
+                schema,
+                {}
+            );
+            mockRequest = {
+                query: {
+                    // complexity should be 2 (1 Query + 1 Scalar)
+                    query: `Query {
+                    scalars: {
+                        num
+                    }
+                `,
+                },
+                ip: '123.456',
+            };
+
+            mockResponse = {
+                json: jest.fn(),
+                send: jest.fn(),
+                sendStatus: jest.fn(),
+                locals: {},
+            };
+
+            complexRequest = {
+                // complexity should be 10 if 'first' is accounted for.
+                // scalars: 1, droid: 1, reviews (4 * (1 Review, 1 episode))
+                query: {
+                    query: `Query {
+                    scalars: {
+                        num
+                    }
+                    droid(id: 1) {
+                        name
+                    }
+                    reviews(episode: 'NEWHOPE', first: 4) {
+                        episode 
+                        stars
+                        commentary
+                    }
+                `,
+                },
+            };
+        });
+
+        xdescribe('Adds expected properties to res.locals', () => {
+            test('Adds UNIX timestamp and complexity', () => {
+                const expectedResponse = {
+                    locals: {},
+                };
+
+                middleware(mockRequest as Request, mockResponse as Response, nextFunction);
+
+                // We don't actually call json
+                expect(mockResponse.json).toBeCalledWith(expectedResponse);
+                expect(mockResponse.locals).toHaveProperty('complexity');
+                expect(mockResponse.locals?.complexity).toBeInstanceOf('number');
+                expect(mockResponse.locals?.complexity).toBeGreaterThanOrEqual(0);
+
+                expect(mockResponse.locals).toHaveProperty('timestamp');
+                expect(mockResponse.locals?.timestamp).toBeInstanceOf('number');
+                // confirm that this is timestamp +/- 5 minutes of now.
+                const now: number = Date.now().valueOf();
+                const diff: number = Math.abs(now - (mockResponse.locals?.timestamp || 0));
+                expect(diff).toBeLessThan(5 * 60);
+            });
+        });
+
+        describe('Correctly limits requests', () => {
+            describe('Allows requests', () => {
+                test('...a single request', () => {
+                    // successful request calls next without any arguments.
+                    middleware(mockRequest as Request, mockResponse as Response, nextFunction);
+                    expect(nextFunction).toBeCalledTimes(1);
+                    expect(nextFunction).toBeCalledWith();
+                });
+
+                test('Multiple valid requests at > 1 second intervals', () => {
```

Comment:
Is anything special happening around the 1 second mark that we need to test above and below 1 second explicitly.